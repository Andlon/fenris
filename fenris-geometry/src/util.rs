use nalgebra::{RealField, UnitVector3, Vector3};

pub fn compute_orthonormal_vectors_3d<T: RealField>(vector: &UnitVector3<T>) -> [UnitVector3<T>; 2] {
    // Ported from
    // https://github.com/dimforge/parry/blob/ac8dcf0197066cd2413a20e4420961b4694996c0/src/utils/wops.rs#L120-L138
    // originally based on the Pixar paper "Building an Orthonormal Basis, Revisited",
    // https://graphics.pixar.com/library/OrthonormalB/paper.pdf
    let v = vector;
    let sign = T::copysign(T::one(), v.z);
    let a = -T::one() / (sign + v.z);
    let b = v.x * v.y * a;

    [
        Vector3::new(T::one() + sign * v.x * v.x * a, sign * b, -sign * v.x),
        Vector3::new(b, sign + v.y * v.y * a, -v.y),
    ]
    .map(UnitVector3::new_unchecked)
}

/// Compares two arrays for *shift-invariant* equality with the given comparator function.
///
/// If `X` and `Y` are the two arrays, then the two arrays are shift-invariant equal if X can be shifted/rotated
/// by some constant `n` such that `Shift(X) == Y`.
pub fn slices_are_equal_shift_invariant<T, C: Fn(&T, &T) -> bool>(x: &[T], y: &[T], comparator: C) -> bool {
    let n = x.len();
    if y.len() != n {
        return false;
    } else if n == 0 {
        // Empty arrays are always equal
        return true;
    }

    'outer_loop: for i_start in 0..n {
        for (j, y_j) in y.iter().enumerate() {
            let x_j_shifted = &x[(j + i_start) % n];
            if !comparator(x_j_shifted, y_j) {
                continue 'outer_loop;
            }
        }
        return true;
    }

    return false;
}

// Base macro for line segment assertions
#[doc(hidden)]
#[macro_export]
macro_rules! assert_line_segments_approx_equal_base {
    ($msg_handler:expr, $segment1:expr, $segment2:expr, abstol = $tol:expr) => {{
        use fenris_geometry::util::slices_are_equal_shift_invariant;
        use matrixcompare::comparators::AbsoluteElementwiseComparator;
        use matrixcompare::compare_matrices;
        use nalgebra::Point2;

        let tol = $tol.clone();
        // Type check: Makes for an easier error message than failing specific methods
        let (segment1, segment2): (&LineSegment2d<_>, &LineSegment2d<_>) = (&$segment1, &$segment2);
        let vertices1 = [segment1.start().clone(), segment1.end().clone()];
        let vertices2 = [segment2.start().clone(), segment2.end().clone()];
        let comparator = AbsoluteElementwiseComparator { tol };
        let comparator = |a: &Point2<f64>, b: &Point2<f64>| compare_matrices(&a.coords, &b.coords, &comparator).is_ok();
        let vertices_are_shift_invariant_equal = slices_are_equal_shift_invariant(&vertices1, &vertices2, comparator);
        if !vertices_are_shift_invariant_equal {
            let msg = format!(
                "Line segments are not (approximately) equal to absolute tolerance {tol}.
Segment1: {:?}
Segment2: {:?}",
                segment1, segment2
            );

            return $msg_handler(msg);
        }
    }};
}

#[macro_export]
macro_rules! assert_line_segments_approx_equal {
    ($segment1:expr, $segment2:expr, abstol = $tol:expr) => {{
        let msg_handler = |msg| panic!("{}", msg);
        $crate::assert_line_segments_approx_equal_base!(msg_handler, $segment1, $segment2, abstol = $tol);
    }};
}
