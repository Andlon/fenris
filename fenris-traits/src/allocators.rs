//! Helper traits for allocator trait bounds.
use nalgebra::allocator::Allocator;
use nalgebra::{DefaultAllocator, DimName, Scalar, U1};

/// An allocator for a single dimension.
pub trait DimAllocator<T: Scalar, D: DimName>:
Allocator<T, D>
+ Allocator<T, D, D>
+ Allocator<T, U1, D>
// Used for various functionality like decompositions
+ Allocator<usize, D>
+ Allocator<(usize, usize), D>
// Provide allocators for built-in types so that we don't need separate traits to use these
// if we already have Allocator<T, D>
+ Allocator<f32, D>
+ Allocator<f64, D>
+ Allocator<i8, D>
+ Allocator<i32, D>
+ Allocator<i64, D>
+ Allocator<u8, D>
+ Allocator<u16, D>
+ Allocator<u32, D>
+ Allocator<u64, D>
+ Allocator<isize, D>
+ Allocator<bool, D>
{}

impl<T, D> DimAllocator<T, D> for DefaultAllocator
where
    T: Scalar,
    D: DimName,
    DefaultAllocator: Allocator<T, D>
        + Allocator<T, D, D>
        + Allocator<T, U1, D>
        + Allocator<usize, D>
        + Allocator<(usize, usize), D>
        + Allocator<f32, D>
        + Allocator<f64, D>
        + Allocator<i8, D>
        + Allocator<i32, D>
        + Allocator<i64, D>
        + Allocator<u8, D>
        + Allocator<u16, D>
        + Allocator<u32, D>
        + Allocator<u64, D>
        + Allocator<isize, D>
        + Allocator<bool, D>,
{
}

/// An allocator for two dimensions.
pub trait BiDimAllocator<T: Scalar, D1: DimName, D2: DimName>:
    DimAllocator<T, D1> + DimAllocator<T, D2> + Allocator<T, D1, D2> + Allocator<T, D2, D1>
{
}

impl<T: Scalar, D1: DimName, D2: DimName> BiDimAllocator<T, D1, D2> for DefaultAllocator where
    DefaultAllocator: DimAllocator<T, D1> + DimAllocator<T, D2> + Allocator<T, D1, D2> + Allocator<T, D2, D1>
{
}

/// An allocator for three dimensions.
pub trait TriDimAllocator<T: Scalar, D1: DimName, D2: DimName, D3: DimName>:
    BiDimAllocator<T, D1, D2> + BiDimAllocator<T, D1, D3> + BiDimAllocator<T, D2, D3>
{
}

impl<T: Scalar, D1: DimName, D2: DimName, D3: DimName> TriDimAllocator<T, D1, D2, D3> for DefaultAllocator where
    DefaultAllocator: BiDimAllocator<T, D1, D2> + BiDimAllocator<T, D1, D3> + BiDimAllocator<T, D2, D3>
{
}

pub trait QuadDimAllocator<T: Scalar, D1: DimName, D2: DimName, D3: DimName, D4: DimName>:
    TriDimAllocator<T, D1, D2, D3>
    + TriDimAllocator<T, D1, D2, D4>
    + TriDimAllocator<T, D1, D3, D4>
    + TriDimAllocator<T, D2, D3, D4>
{
}

impl<T: Scalar, D1: DimName, D2: DimName, D3: DimName, D4: DimName> QuadDimAllocator<T, D1, D2, D3, D4>
    for DefaultAllocator
where
    DefaultAllocator: TriDimAllocator<T, D1, D2, D3>
        + TriDimAllocator<T, D1, D2, D4>
        + TriDimAllocator<T, D1, D3, D4>
        + TriDimAllocator<T, D2, D3, D4>,
{
}
