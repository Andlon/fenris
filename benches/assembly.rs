use criterion::{criterion_group, criterion_main, Criterion};
use fenris::assembly::global::CsrAssembler;
use fenris::assembly::local::{ElementEllipticAssemblerBuilder, QuadratureTable};
use fenris::assembly::operators::LaplaceOperator;
use fenris::element::ElementConnectivity;
use fenris::mesh::procedural::create_unit_box_uniform_tet_mesh_3d;
use fenris::mesh::Mesh;
use fenris::quadrature::CanonicalStiffnessQuadrature;
use fenris::SmallDim;
use fenris_traits::allocators::DimAllocator;
use nalgebra::{DVector, DVectorSlice, DefaultAllocator};
use nalgebra_sparse::pattern::SparsityPattern;
use nalgebra_sparse::CsrMatrix;
use std::hint::black_box;
use fenris_solid::MaterialEllipticOperator;
use fenris_solid::materials::{LameParameters, LinearElasticMaterial};

fn assemble_poisson_into_serial<D, C>(
    matrix: &mut CsrMatrix<f64>,
    assembler: &CsrAssembler<f64>,
    u: DVectorSlice<f64>,
    qtable: &impl QuadratureTable<f64, D, Data = ()>,
    mesh: &Mesh<f64, D, C>,
) -> eyre::Result<()>
where
    D: SmallDim,
    C: ElementConnectivity<f64, GeometryDim = D, ReferenceDim = D>,
    DefaultAllocator: DimAllocator<f64, D>,
{
    let element_assembler = ElementEllipticAssemblerBuilder::new()
        .with_u(u)
        .with_finite_element_space(mesh)
        .with_operator(&LaplaceOperator)
        .with_quadrature_table(qtable)
        .build();
    assembler.assemble_into_csr(matrix, &element_assembler)
}

fn assemble_poisson_pattern_serial<D, C>(
    assembler: &CsrAssembler<f64>,
    u: DVectorSlice<f64>,
    qtable: &impl QuadratureTable<f64, D, Data = ()>,
    mesh: &Mesh<f64, D, C>,
) -> SparsityPattern
where
    D: SmallDim,
    C: ElementConnectivity<f64, GeometryDim = D, ReferenceDim = D>,
    DefaultAllocator: DimAllocator<f64, D>,
{
    let element_assembler = ElementEllipticAssemblerBuilder::new()
        .with_u(u)
        .with_finite_element_space(mesh)
        .with_operator(&LaplaceOperator)
        .with_quadrature_table(qtable)
        .build();
    assembler.assemble_pattern(&element_assembler)
}

fn assemble_elasticity_pattern_serial<D, C>(
    assembler: &CsrAssembler<f64>,
    u: DVectorSlice<f64>,
    qtable: &impl QuadratureTable<f64, D, Data = LameParameters<f64>>,
    mesh: &Mesh<f64, D, C>,
) -> SparsityPattern
    where
        D: SmallDim,
        C: ElementConnectivity<f64, GeometryDim = D, ReferenceDim = D>,
        DefaultAllocator: DimAllocator<f64, D>,
{
    let material = LinearElasticMaterial;
    let operator = MaterialEllipticOperator::new(&material);
    let element_assembler = ElementEllipticAssemblerBuilder::new()
        .with_u(u)
        .with_finite_element_space(mesh)
        .with_operator(&operator)
        .with_quadrature_table(qtable)
        .build();
    assembler.assemble_pattern(&element_assembler)
}

pub fn poisson_assembly_serial(c: &mut Criterion) {
    let resolutions = vec![5, 10, 20];
    let assembler = CsrAssembler::default();
    for res in resolutions {
        let tet4_mesh = create_unit_box_uniform_tet_mesh_3d(res);
        let pattern = assembler.assemble_pattern(&tet4_mesh);
        let nnz = pattern.nnz();
        let mut matrix = CsrMatrix::try_from_pattern_and_values(pattern, vec![0.0; nnz]).unwrap();
        let u = DVector::repeat(matrix.nrows(), 0.0);
        let qtable = tet4_mesh.canonical_stiffness_quadrature();
        c.bench_function(
            &format!("serial assembly poisson stiffness matrix tet4 (res={res})"),
            |b| {
                b.iter(|| {
                    assemble_poisson_into_serial(&mut matrix, &assembler, DVectorSlice::from(&u), &qtable, &tet4_mesh)
                })
            },
        );
    }
}

pub fn poisson_pattern_assembly_serial(c: &mut Criterion) {
    let resolutions = vec![5, 10, 20];
    let assembler = CsrAssembler::default();
    for res in resolutions {
        let tet4_mesh = create_unit_box_uniform_tet_mesh_3d(res);
        let u = DVector::repeat(tet4_mesh.vertices().len(), 0.0);
        let qtable = tet4_mesh.canonical_stiffness_quadrature();
        c.bench_function(
            &format!("serial pattern assembly poisson stiffness matrix tet4 (res={res})"),
            |b| {
                b.iter(|| {
                    black_box(assemble_poisson_pattern_serial(
                        &assembler,
                        DVectorSlice::from(&u),
                        &qtable,
                        &tet4_mesh,
                    ))
                })
            },
        );
    }
}

pub fn elasticity_3d_pattern_assembly_serial(c: &mut Criterion) {
    let resolutions = vec![5, 10, 20];
    let assembler = CsrAssembler::default();
    for res in resolutions {
        let tet4_mesh = create_unit_box_uniform_tet_mesh_3d(res);
        let u = DVector::repeat(tet4_mesh.vertices().len(), 0.0);
        let qtable = tet4_mesh.canonical_stiffness_quadrature()
            .with_uniform_data(LameParameters::default());
        c.bench_function(
            &format!("serial pattern assembly elasticity stiffness matrix tet4 (res={res})"),
            |b| {
                b.iter(|| {
                    black_box(assemble_elasticity_pattern_serial(
                        &assembler,
                        DVectorSlice::from(&u),
                        &qtable,
                        &tet4_mesh,
                    ))
                })
            },
        );
    }
}

criterion_group!(
    serial_assembly,
    poisson_assembly_serial,
    poisson_pattern_assembly_serial,
    elasticity_3d_pattern_assembly_serial,
);

criterion_main!(serial_assembly);
