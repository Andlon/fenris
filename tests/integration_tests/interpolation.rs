use itertools::{izip, Itertools};
use matrixcompare::assert_matrix_eq;
use nalgebra::{DefaultAllocator, DVectorSlice, OMatrix, OVector, Point2, U1, U2, vector, Vector1, Vector2};
use fenris::assembly::buffers::{BufferUpdate, InterpolationBuffer};
use fenris::io::vtk::FiniteElementMeshDataSetBuilder;
use fenris::mesh::procedural::create_unit_square_uniform_tri_mesh_2d;
use fenris::mesh::{Mesh, TriangleMesh2d};
use fenris::space::{InterpolateGradientInSpace, InterpolateInSpace, SpatiallyIndexed};
use fenris::util::global_vector_from_point_fn;
use fenris::{quadrature, SmallDim};
use fenris::connectivity::Tri3d2Connectivity;
use fenris::mesh::refinement::refine_uniformly_repeat;
use fenris_traits::allocators::{BiDimAllocator, TriDimAllocator};
use util::flatten_vertically;
use crate::integration_tests::data_output_path;

fn u_scalar(p: &Point2<f64>) -> Vector1<f64> {
    let (x, y) = (p.x, p.y);
    Vector1::new((x.cos() + y.sin()) * x.powi(2))
}

fn u_vector(p: &Point2<f64>) -> Vector2<f64> {
    let (x, y) = (p.x, p.y);
    vector![(x.cos() + y.sin()) * x.powi(2),
            (x.powi(2) + 0.5).ln() * (y.powi(2) + 0.25).ln() + x * y + 3.0]
}

struct ExpectedInterpolationTestValues<SolutionDim>
where
    SolutionDim: SmallDim,
    DefaultAllocator: BiDimAllocator<f64, U2, SolutionDim>,
{
    u_expected: Vec<OVector<f64, SolutionDim>>,
    grad_u_expected: Vec<OMatrix<f64, U2, SolutionDim>>,
    u_interpolated: Vec<OVector<f64, SolutionDim>>,
    grad_u_interpolated: Vec<OMatrix<f64, U2, SolutionDim>>,
}

fn compute_expected_interpolation_test_values<'a, Space, SolutionDim>(
    space: &Space,
    quadrature_points: &[Point2<f64>],
    u_vec: impl Into<DVectorSlice<'a, f64>>,
) -> ExpectedInterpolationTestValues<SolutionDim>
where
    SolutionDim: SmallDim,
    Space: InterpolateGradientInSpace<f64, SolutionDim, GeometryDim=U2, ReferenceDim=U2>,
    DefaultAllocator: TriDimAllocator<f64, U2, U2, SolutionDim>
{
    let u_vec = u_vec.into();
    let mut interpolation_buffer = InterpolationBuffer::default();

    // For each element, compute interpolated value + gradient at quadrature points plus
    // the map to physical space. Then later we'll interpolate at these same points (in physical
    // space), so that we already know the correct answer.
    let (x_expected, u_expected, grad_u_expected): (Vec<_>, Vec<_>, Vec<_>) = (0 .. space.num_elements())
        .flat_map(|i| {
            let mut buffer = interpolation_buffer.prepare_element_in_space(i, space, u_vec, SolutionDim::dim());
            quadrature_points
                .iter()
                .map(|xi_j| {
                    buffer.update_reference_point(xi_j, BufferUpdate::Both);
                    let u_j: OVector<_, SolutionDim> = buffer.interpolate();
                    let grad_u_j_ref: OMatrix<_, Space::ReferenceDim, SolutionDim> = buffer.interpolate_ref_gradient();
                    let j_inv_t = buffer.element_reference_jacobian()
                        .try_inverse()
                        .unwrap()
                        .transpose();
                    let grad_u_j = j_inv_t * grad_u_j_ref;
                    let x_j = space.map_element_reference_coords(i, xi_j);
                    (x_j, u_j, grad_u_j)
                }).collect::<Vec<_>>()
        })
        .multiunzip();

    let mut u_interpolated = vec![OVector::<_, SolutionDim>::zeros(); x_expected.len()];
    let mut grad_u_interpolated = vec![OMatrix::<_, U2, SolutionDim>::zeros(); x_expected.len()];
    space.interpolate_at_points(&x_expected, DVectorSlice::from(&u_vec), &mut u_interpolated);
    space.interpolate_gradient_at_points(&x_expected, DVectorSlice::from(&u_vec), &mut grad_u_interpolated);

    ExpectedInterpolationTestValues {
        u_expected,
        grad_u_expected,
        u_interpolated,
        grad_u_interpolated,
    }
}

#[test]
fn spatially_indexed_interpolation_trimesh() {
    // We interpolate at (quadrature) points of a finite element space
    // in two ways:
    //  - by computing the values in reference coordinate space of each element
    //    (this forms the "expected" values)
    //  - by interpolating the quantity at the *physical* coordinates
    // This way we verify that the latter approach produces expected results.
    let mesh: TriangleMesh2d<f64> = create_unit_square_uniform_tri_mesh_2d(10);

    // Arbitrary scalar function u(p), where p is a 2-dimensional point
    let u_weights_scalar = global_vector_from_point_fn(mesh.vertices(), u_scalar);
    let u_weights_vector = global_vector_from_point_fn(mesh.vertices(), u_vector);
    let space = SpatiallyIndexed::from_space(mesh);

    let (_, interior_points) = quadrature::total_order::triangle::<f64>(4).unwrap();
    let interface_points = [
        // Points on the boundary of the reference element, which will be mapped to
        // the boundary of the physical element, and thus on an interface between
        // neighboring elements
        [-1.0, -1.0],
        [1.0, -1.0],
        [-1.0, 1.0],
        [-1.0, 0.5],
        [0.5, -1.0],
        [0.0, 0.0],
    ].map(Point2::from);

    // For debugging
    FiniteElementMeshDataSetBuilder::from_mesh(space.space())
        .try_export(data_output_path()
            .join("interpolation/spatially_indexed_interpolation_trimesh/mesh.vtu"))
        .unwrap();

    // For each element, compute interpolated value of quadrature points plus
    // the map to physical space. Then later we'll interpolate at these same points (in physical
    // space), so that we already know the correct answer.
    {
        // For interior quadrature points, we check both values and gradients of the scalar function
        let values = compute_expected_interpolation_test_values::<_, U1>(&space, &interior_points, &u_weights_scalar);
        let iter = izip!(values.u_interpolated, values.grad_u_interpolated, values.u_expected, values.grad_u_expected);
        for (u, grad_u, u_expected, grad_u_expected) in iter {
            assert_matrix_eq!(u, u_expected, comp = abs, tol = 1e-12);
            assert_matrix_eq!(grad_u, grad_u_expected, comp = abs, tol = 1e-12);
        }
    }

    {
        // For boundary points, we only check values since gradients are discontinuous
        // at element interfaces
        let values = compute_expected_interpolation_test_values::<_, U1>(&space, &interface_points, &u_weights_scalar);
        let iter = izip!(values.u_interpolated, values.u_expected);
        for (u, u_expected) in iter {
            assert_matrix_eq!(u, u_expected, comp = abs, tol = 1e-12);
        }
    }

    {
        // Repeat interior quadrature points for vector function
        let values = compute_expected_interpolation_test_values::<_, U2>(&space, &interior_points, &u_weights_vector);
        let iter = izip!(values.u_interpolated, values.grad_u_interpolated, values.u_expected, values.grad_u_expected);
        for (u, grad_u, u_expected, grad_u_expected) in iter {
            assert_matrix_eq!(u, u_expected, comp = abs, tol = 1e-12);
            assert_matrix_eq!(grad_u, grad_u_expected, comp = abs, tol = 1e-12);
        }
    }

    {
        // Repeat interface quadrature points for vector function
        let values = compute_expected_interpolation_test_values::<_, U2>(&space, &interface_points, &u_weights_vector);
        let iter = izip!(values.u_interpolated, values.u_expected);
        for (u, u_expected) in iter {
            assert_matrix_eq!(u, u_expected, comp = abs, tol = 1e-12);
        }
    }
}

#[test]
fn basic_extrapolation() {
    // We don't have any guarantees about how extrapolation should perform,
    // but we want to make sure it doesn't do anything crazy. Therefore we use this test
    // to visually confirm that we get reasonable behavior for a simple example,
    // and use insta to track any changes in its output.

    // The set up consists of two meshes of a square with a hole. The "base" mesh is smaller,
    // and the "outer" mesh is slightly thicker, so that there are parts of "outer" that
    // extend beyond the domain of "base". We then interpolate an arbitrary function
    // from "base" onto "outer", so that some parts are interpolated and some are extrapolated.

    // Increase this for higher fidelity visualization
    let refinement_rounds = 2;

    // The s variable determines the "extruded" thickness
    let vertices = |s: f64| [
        [-s, -s],
        [1.0, -s],
        [2.0, -s],
        [3.0 + s, -s],
        [-s, 1.0],
        [1.0 + s, 1.0 + s],
        [2.0 - s, 1.0 + s],
        [3.0 + s, 1.0 + s],
        [0.0 - s, 2.0 - s],
        [1.0 + s, 2.0 - s],
        [2.0 - s, 2.0 - s],
        [3.0 + s, 2.0 - s],
        [0.0 - s, 3.0 + s],
        [1.0, 3.0 + s],
        [2.0, 3.0 + s],
        [3.0 + s, 3.0 + s],
    ].map(Point2::from);
    let connectivity = [
        [0, 1, 4],
        [1, 5, 4],
        [1, 2, 6],
        [1, 5, 6],
        [2, 3, 6],
        [3, 7, 6],
        [6, 7, 11],
        [6, 11, 10],
        [10, 11, 14],
        [11, 15, 14],
        [10, 14, 9],
        [9, 14, 13],
        [12, 9, 13],
        [8, 9, 12],
        [4, 9, 8],
        [4, 5, 9],
    ].map(Tri3d2Connectivity);
    let base_mesh = Mesh::from_vertices_and_connectivity(vertices(0.0).to_vec(), connectivity.to_vec());
    let base_mesh = refine_uniformly_repeat(&base_mesh, refinement_rounds);
    let outer_mesh = Mesh::from_vertices_and_connectivity(vertices(0.1).to_vec(), connectivity.to_vec());
    let outer_mesh = refine_uniformly_repeat(&outer_mesh, refinement_rounds);

    let u_base = global_vector_from_point_fn(base_mesh.vertices(), u_scalar);
    FiniteElementMeshDataSetBuilder::from_mesh(&base_mesh)
        .with_point_scalar_attributes("u", 1, u_base.as_slice())
        .try_export(data_output_path()
            .join("interpolation/extrapolation/base_mesh.vtu"))
        .unwrap();

    let mut u_outer = vec![Vector1::zeros(); outer_mesh.vertices().len()];
    SpatiallyIndexed::from_space(base_mesh)
        .interpolate_at_points(outer_mesh.vertices(), DVectorSlice::from(&u_base), &mut u_outer);

    FiniteElementMeshDataSetBuilder::from_mesh(&outer_mesh)
        .with_point_scalar_attributes("u", 1, flatten_vertically(&u_outer).unwrap().as_slice())
        .try_export(data_output_path()
            .join("interpolation/extrapolation/outer_mesh.vtu"))
        .unwrap();

    insta::assert_debug_snapshot!(&u_outer);
}