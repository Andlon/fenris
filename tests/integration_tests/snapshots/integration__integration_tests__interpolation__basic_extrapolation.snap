---
source: tests/integration_tests/interpolation.rs
expression: "&u_outer"
---
[
    [
        [
            0.0,
        ],
    ],
    [
        [
            0.042389918449840705,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            0.1876279176494576,
        ],
    ],
    [
        [
            0.04857501743120377,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            0.3923570539146311,
        ],
    ],
    [
        [
            0.22164596204695453,
        ],
    ],
    [
        [
            0.5403023058681398,
        ],
    ],
    [
        [
            0.4820409891443957,
        ],
    ],
    [
        [
            0.2779047656353667,
        ],
    ],
    [
        [
            0.06046451602190358,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            0.070947781566992,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            0.7382254732717581,
        ],
    ],
    [
        [
            1.0641409946402092,
        ],
    ],
    [
        [
            0.668200640367969,
        ],
    ],
    [
        [
            1.368328200669842,
        ],
    ],
    [
        [
            0.8466817813624639,
        ],
    ],
    [
        [
            1.5520598164076003,
        ],
    ],
    [
        [
            1.0118269611448256,
        ],
    ],
    [
        [
            0.4351390684019518,
        ],
    ],
    [
        [
            0.36114074231642135,
        ],
    ],
    [
        [
            0.09119113054185013,
        ],
    ],
    [
        [
            0.4926911912426073,
        ],
    ],
    [
        [
            0.8067072517733143,
        ],
    ],
    [
        [
            0.15915870375233154,
        ],
    ],
    [
        [
            0.6378390791595003,
        ],
    ],
    [
        [
            1.2386516515517656,
        ],
    ],
    [
        [
            -0.5458785454265696,
        ],
    ],
    [
        [
            0.13076487966490158,
        ],
    ],
    [
        [
            -1.6645873461885696,
        ],
    ],
    [
        [
            -0.7610237964978965,
        ],
    ],
    [
        [
            0.3869642798923545,
        ],
    ],
    [
        [
            0.9854629663613995,
        ],
    ],
    [
        [
            1.587001884142874,
        ],
    ],
    [
        [
            1.3337352127044546,
        ],
    ],
    [
        [
            1.8332284940448553,
        ],
    ],
    [
        [
            1.2417935953116745,
        ],
    ],
    [
        [
            1.6319592536477217,
        ],
    ],
    [
        [
            1.8564853679179765,
        ],
    ],
    [
        [
            2.052468419570099,
        ],
    ],
    [
        [
            1.7647704149578864,
        ],
    ],
    [
        [
            2.0353947603521108,
        ],
    ],
    [
        [
            -3.362830828247313,
        ],
    ],
    [
        [
            -5.403725425416268,
        ],
    ],
    [
        [
            -2.178142930053049,
        ],
    ],
    [
        [
            -7.566005457706808,
        ],
    ],
    [
        [
            -3.968416715019687,
        ],
    ],
    [
        [
            -8.909932469404009,
        ],
    ],
    [
        [
            -5.877221931117341,
        ],
    ],
    [
        [
            -2.010737980892067,
        ],
    ],
    [
        [
            -0.6524219774422563,
        ],
    ],
    [
        [
            0.5567931246760125,
        ],
    ],
    [
        [
            -7.128623962771442,
        ],
    ],
    [
        [
            -4.5951026219661815,
        ],
    ],
    [
        [
            -3.6105254065642116,
        ],
    ],
    [
        [
            -2.487485624581789,
        ],
    ],
    [
        [
            -1.781405378437382,
        ],
    ],
    [
        [
            -0.949644521761257,
        ],
    ],
    [
        [
            -0.4123297394259959,
        ],
    ],
    [
        [
            0.520830144250242,
        ],
    ],
    [
        [
            -0.5471150660417942,
        ],
    ],
    [
        [
            1.421828745012562,
        ],
    ],
    [
        [
            1.7746715094488597,
        ],
    ],
    [
        [
            0.9846442328646636,
        ],
    ],
    [
        [
            1.2271960691070047,
        ],
    ],
    [
        [
            0.16283203838138863,
        ],
    ],
    [
        [
            -0.28175223415649,
        ],
    ],
    [
        [
            0.06752241003248127,
        ],
    ],
    [
        [
            0.45631976023462145,
        ],
    ],
    [
        [
            0.37070622094500166,
        ],
    ],
    [
        [
            -0.029742676024364867,
        ],
    ],
    [
        [
            -0.4573769557991544,
        ],
    ],
    [
        [
            2.17015942503255,
        ],
    ],
    [
        [
            1.9608300437648563,
        ],
    ],
    [
        [
            2.325392600227648,
        ],
    ],
    [
        [
            1.159650870456417,
        ],
    ],
    [
        [
            -0.010098310879632214,
        ],
    ],
    [
        [
            0.8626826206128133,
        ],
    ],
    [
        [
            1.9061184328578802,
        ],
    ],
    [
        [
            2.282163673091272,
        ],
    ],
    [
        [
            1.6684448214272112,
        ],
    ],
    [
        [
            2.079099346359327,
        ],
    ],
    [
        [
            1.6448598358038657,
        ],
    ],
    [
        [
            0.960615011738404,
        ],
    ],
    [
        [
            0.8408294633601648,
        ],
    ],
    [
        [
            -0.059103456049724036,
        ],
    ],
    [
        [
            -1.110079226196019,
        ],
    ],
    [
        [
            -1.5061755869392517,
        ],
    ],
    [
        [
            -0.15036373550758683,
        ],
    ],
    [
        [
            -0.25428600310192023,
        ],
    ],
    [
        [
            -1.628897559468539,
        ],
    ],
    [
        [
            -1.1001073139491009,
        ],
    ],
    [
        [
            -1.6710700835247496,
        ],
    ],
    [
        [
            -3.5236831724683997,
        ],
    ],
    [
        [
            -2.7819687557758046,
        ],
    ],
    [
        [
            -5.907957312119457,
        ],
    ],
    [
        [
            -4.8789211407770825,
        ],
    ],
    [
        [
            -7.639852396865203,
        ],
    ],
    [
        [
            -6.437927393278244,
        ],
    ],
    [
        [
            -4.484681372926383,
        ],
    ],
    [
        [
            -3.218834444666038,
        ],
    ],
    [
        [
            -2.6316527864871264,
        ],
    ],
    [
        [
            2.232091442203688,
        ],
    ],
    [
        [
            1.9139687698678705,
        ],
    ],
    [
        [
            2.2050779141101153,
        ],
    ],
    [
        [
            1.2869423958917974,
        ],
    ],
    [
        [
            0.3769860107177252,
        ],
    ],
    [
        [
            1.4619653015673464,
        ],
    ],
    [
        [
            1.9615890225513997,
        ],
    ],
    [
        [
            1.9717903193482114,
        ],
    ],
    [
        [
            1.8098559838465853,
        ],
    ],
    [
        [
            1.6351472078793872,
        ],
    ],
    [
        [
            1.4616368663307902,
        ],
    ],
    [
        [
            1.4433873387212783,
        ],
    ],
    [
        [
            1.196580343258685,
        ],
    ],
    [
        [
            0.8506147688104435,
        ],
    ],
    [
        [
            -0.11369852074322617,
        ],
    ],
    [
        [
            0.47667872188703275,
        ],
    ],
    [
        [
            0.990216185631819,
        ],
    ],
    [
        [
            0.877031990112793,
        ],
    ],
    [
        [
            0.7131912038361499,
        ],
    ],
    [
        [
            0.681422313928007,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            0.06752867068814881,
        ],
    ],
    [
        [
            0.048563918802459895,
        ],
    ],
    [
        [
            0.3690136764985823,
        ],
    ],
    [
        [
            0.2842539777233626,
        ],
    ],
    [
        [
            0.2176159193621794,
        ],
    ],
    [
        [
            0.9693128761315483,
        ],
    ],
    [
        [
            0.7872714578172095,
        ],
    ],
    [
        [
            0.5991984360718087,
        ],
    ],
    [
        [
            0.4673270581964356,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            0.09484409892741953,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            0.45418884918383917,
        ],
    ],
    [
        [
            0.0889895838041503,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            1.045168696614398,
        ],
    ],
    [
        [
            0.42047515119146145,
        ],
    ],
    [
        [
            0.07836922829073006,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            0.09589485205331155,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            0.4663438687611834,
        ],
    ],
    [
        [
            0.09832037041573496,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            1.0791909680893639,
        ],
    ],
    [
        [
            0.4664053051708363,
        ],
    ],
    [
        [
            0.09772934992754233,
        ],
    ],
    [
        [
            0.0,
        ],
    ],
    [
        [
            0.45664179531148974,
        ],
    ],
    [
        [
            1.0590114914728255,
        ],
    ],
    [
        [
            1.4989889986734197,
        ],
    ],
    [
        [
            1.5377972924721943,
        ],
    ],
    [
        [
            1.0829634853017567,
        ],
    ],
    [
        [
            1.5269900606881004,
        ],
    ],
]
